tempesta_fw/apm.c

```diff
@@ -0,0 +1,1039 @@
+ * Readers read the latest percentiles at @prcntl[@rdidx % 2]. The writer
+ * writes the new percentiles to @prcntl[(@rdidx + 1) % 2], and then
+ * increments @rdidx. The reading and writing is protected by a rwlock.
+ * @prcntl	- The percentiles for reading/writing (flip-flop manner).
+ * @rdidx	- The current index in @prcntl for readers.
+ */
+typedef struct {
+	TfwApmPrcntl	prcntl[2];
+	atomic_t	rdidx;
+} TfwApmStats;
+
+/*
+ * The ring buffer list entry structure.
+ * This is a supporting structure used in the calculation of percentiles.
+ * @ctl		- A copy of response time ranges in the ring buffer entry.
+ * @pcntrng	- A pointer to the stats data for the ring buffer entry.
+ */
+typedef struct {
+	TfwPcntCtl	ctl[TFW_STATS_RANGES];
+	TfwPcntRanges	*pcntrng;
+} TfwApmRBLstEnt;
+
+/*
+ * The ring buffer list structure. 
+ * This is a supporting structure used in the calculation of percentiles.
+ * The list consists of entries of the ring buffer that take part in the
+ * calculation of percentiles. Also it keeps related data that are useful
+ * in making decisions on the recalculation of percentiles.
+ * @rblstent	- The array of ring buffer entries for the calculation.
+ * @rblstsz	- The current size of @rblstent.
+ * @rblstmaxsz	- The maximum size of @rblstent.
+ * @jtmwstamp	- The start of the time window the percentiles are for.
+ * @entry_cnt	- The number of hits in the current buffer ring entry.
+ * @total_cnt	- The number of hits within the current time window.
+ */
+typedef struct {
+	TfwApmRBLstEnt	*rblstent;
+	unsigned int	rblstsz;
+	unsigned int	rblstmaxsz;
+	unsigned long	jtmwstamp;
+	unsigned long	entry_cnt;
+	unsigned long	total_cnt;
+} TfwApmRBLst;
+
+/*
+ * APM Data structure.
+ * Note that the organization of the supporting data heavily depends
+ * on the fact that there's only one party that does the calculation
+ * of percentiles - the function that runs periodically on timer.
+ * If there are several different parties that do the calculation,
+ * then the data may need to be organized differently.
+ * @rbuf	- The ring buffer for the specified time window.
+ * @rblst	- The list of ring buffer entries used in the calculation.
+ * @stats	- The latest percentiles.
+ * @timer	- The periodic timer handle.
+ * @flags	- The atomic flags (see below).
+ */
+#define TFW_APM_DATA_F_REARM	0x0001		/* Re-arm the timer. */
+#define TFW_APM_TIMER_TIMEOUT	(HZ * 3)	/* The timer periodicity. */
+
+typedef struct {
+	TfwApmRBuf		rbuf;
+	TfwApmRBLst		rblst;
+	TfwApmStats		stats;
+	struct timer_list	timer;
+	unsigned long		flags;
+} TfwApmData;
+
+/*
+ * [1ms, 349ms] should be sufficient for almost any installation,
+ * including cross atlantic.
+ */
+static const TfwPcntCtl __read_mostly tfw_rngctl_init[TFW_STATS_RANGES] = {
+	{{0, 1, 16}},
+	{{1, 17, 47}},
+	{{2, 48, 108}},
+	{{4, 109, 349}}
+};
+
+/* A superset of percentiles for all users. */
+const unsigned int __read_mostly tfw_apm_prcntl_ith[] = {
+	1, 50, 75, 90, 95, 99
```

Comment:
Oh, missed the line previous time. I used 1st percentilie for test only. In real life I don't thing that anybody is interested in 1st percentile :)